src/Google/Http/REST.php

```diff
@@ -107,44 +107,70 @@ public static function decodeHttpResponse(
       RequestInterface $request = null,
       $expectedClass = null
   ) {
-    $body = (string) $response->getBody();
     $code = $response->getStatusCode();
-    $result = null;
+    $body = self::decodeBody($response, $request);
 
-    // return raw response when "alt" is "media"
-    $isJson = !($request && 'media' == $request->getUri()->getQuery('alt'));
+    // retry strategy
+    if ((intVal($code)) >= 400) {
+      // Check if we received errors, and add those to the Exception for convenience
+      throw new Google_Service_Exception($body, $code, null, self::getResponseErrors($body));
+    }
 
-    // set the result to the body if it's not set to anything else
-    if ($isJson) {
-      $result = json_decode($body, true);
-      if (null === $result && 0 !== json_last_error()) {
-        // in the event of a parse error, return the raw string
-        $result = $body;
-      }
-    } else {
-      $result = $body;
+    if ($expectedClass = self::determineExpectedClass($expectedClass, $request)) {
+      $json = json_decode($body, true);
+
+      return new $expectedClass($json);
     }
 
-    // retry strategy
-    if ((intVal($code)) >= 400) {
-      $errors = null;
-      // Specific check for APIs which don't return error details, such as Blogger.
-      if (isset($result['error']['errors'])) {
-        $errors = $result['error']['errors'];
-      }
-      throw new Google_Service_Exception($body, $code, null, $errors);
+    return $response;
+  }
+
+  private static function decodeBody(ResponseInterface $response, RequestInterface $request = null)
+  {
+    if (self::isAltMedia($request)) {
```

Comment:
This is true. We can probably assume if it's an error, it's safe to pull the body into memory.